### ISSUE TITLE
printf_i executes at k-time

### ISSUE THREAD

**Issue by joachimheintz** [csound/csound#560]
this is an example code:

<CsoundSynthesizer>
<CsOptions>
-odac
</CsOptions>
<CsInstruments>
ksmps = 32

instr One

 S_arr[] init 4
 S_arr[0] = "ich ging"
 S_arr[1] = "im dunklen"
 S_arr[2] = "so fuer"
 S_arr[3] = "mich hin"
  
 indx = 0
 while indx < 4 do
  printf_i "  S_arr[%d] = '%s'\n", 1, indx, S_arr[indx]
  indx += 1
 od
 
endin

</CsInstruments>
<CsScore>
i "One" 0 .1
</CsScore>
</CsoundSynthesizer>

it leads to this output:

new alloc for instr One:
  S_arr[0] = 'ich ging'
  S_arr[1] = 'im dunklen'
  S_arr[2] = 'so fuer'
  S_arr[3] = 'mich hin'
WARNING: Array index 4 out of range (0,3) for dimension 1
WARNING: Array index 4 out of range (0,3) for dimension 1
WARNING: Array index 4 out of range (0,3) for dimension 1
...

array index is indeed 4, but only at the end of the while loop.  it seems to me as if this expression is then executed at perf-time.  this should not happen, as printf_i should only execute at i-time.

**Comment by jpffitch** [comment]
The loop is an i-rate loop so it looks correct to me -- all at init time

**Comment by joachimheintz** [comment]
yes so printf_i should do nothing, right?

**Comment by jpffitch** [comment]
No; it shoud print once in every time round the loop and give an  error if 
indx too large -- that bit I habve not understood yet

**Comment by joachimheintz** [comment]
yes, of course: it should print the content of the indices 0..3, and then do nothing more, because index=4 is not applied anywhere in the code.  but it is obviously applied at perf-time by csound.

**Comment by kunstmusik** [comment]
I suspect the issue is not the printf_i but the array access.  It seems to
me that it is doing an i-rate and k-rate access of the S_arr array. At
perf-time, the init-time while check is probably being skipped, and the
S_arr array_get is using the indx from init-time (now set to 4).

A work around is to use turnoff when complete with this instrument to
ensure it is init-time only. This kind of code though seems like something
that could or would come up often, so it seems like we need some work to
figure out how to clearly define and understand the array access and when
it operates.

**Comment by vlazzarini** [comment]
printf_i cannot run at perf-time, it does not have a perf-time function. However,
as @kunstmusik said, the array access function seems to be running at perf-time.
It does not make sense to me, as it is an argument to an i-time function. All expressions
in printf_i should only run at i-time. I think this is a bug.

Of course if you run the code with p3 =0, then the perf-time is skipped and no warnings are printed.

**Comment by joachimheintz** [comment]
yes - when i remove printf_i the error disappears:

 indx = 0
 while indx < 4 do
  String = S_arr[indx]
  indx += 1
 od

as you say, it makes no sense that something in printf_i executes at perf-time.

**Comment by vlazzarini** [comment]
printf_i is not executing at perf-time, it can't. But the array_get opcode is, and that is the bug

**Comment by joachimheintz** [comment]
ok, i understand.

**Comment by vlazzarini** [comment]
Ok, so I looked at this with -v and the generated code uses ##array_get.i2  which has a perf-time subroutine, whereas ##array_get.i  should probably have been used (which doesn't). 
This is the reason for the perf_time warnings.

**Comment by vlazzarini** [comment]
There is a solution: to insert a OENTRY in arrays.c with these contents

{ "##array_get.i", sizeof(ARRAY_GET), 0, 1, "S", "S[]m", (SUBR)array_get },

this will match S and S[] and will generate ##array_get.i instead of ##array_get.i2  which is
a sort of default, catch-all. How does that sound @kunstmusik? I tested it and it fixes the problem.

**Comment by kunstmusik** [comment]
This is sort of an ongoing issue to figure.  The issue is that, sometime
you want an i-time argument but perf-time access.  I think we set it such
that the rate of performance depends on the type of the array and not the
index variable.  We should probably revisit this.

**Comment by vlazzarini** [comment]
Should I commit my fix?

**Comment by kunstmusik** [comment]
@vlazzarini the problem is, what if someone is changing an S-type array at
perf-time, but uses constants as indexes?  i.e.

printf "Test %s %s\n", S_arr[0] , S_arr[1]

If the above is wrapped in some kind of k-rate conditional, maybe the user
might expected to see updated values then.

**Comment by vlazzarini** [comment]
That does not seem to be a problem. The values appear to be updated OK:

```
ksmps = 32

instr One

S_arr[] init 4
S_arr[0] = "ich ging"
S_arr[1] = "im dunklen"
S_arr[2] = "so fuer"
S_arr[3] = "mich hin"
k1 init 0
if k1 == 1 then
S_arr[0] = "ich ging2"
S_arr[1] = "im dunklen2"
S_arr[2] = "so fuer2"
S_arr[3] = "mich hin2"
printf "Test %s %s\n",1, S_arr[0] , S_arr[1]
endif
k1 = 1
endin
schedule("One", 0, 0.1)
```

prints
new alloc for instr One:
Test ich ging2 im dunklen2

**Comment by vlazzarini** [comment]
I don't think it is a problem:

```
ksmps = 32

instr One

S_arr[] init 4
S_arr[0] = "ich ging"
S_arr[1] = "im dunklen"
S_arr[2] = "so fuer"
S_arr[3] = "mich hin"
k1 init 0
if k1 == 1 then
S_arr[0] strcpyk "ich ging2"
S_arr[1] strcpyk "im dunklen2"
S_arr[2] strcpyk "so fuer2"
S_arr[3] strcpyk "mich hin2"
printf "Test %s %s\n",1, S_arr[0] , S_arr[1]
endif
k1 = 1
endin
schedule("One", 0, 0.1)
```
prints
Test ich ging2 im dunklen2

**Comment by vlazzarini** [comment]
In this case I think we need to force the use of k-rate indexes. 

k1 = 0
printf "Test %s %s\n", S_arr[k1] , S_arr[k1+1]

I think the problem of having code that is supposed to be i-time being performed at k-time 
is worse.

**Comment by joachimheintz** [comment]
when you discuss this further, maybe it is useful for you to have 
another example.  i actually came across this issue because i wrote a 
UDO which transforms a string to a string array, using any character as 
seperator.  i paste the code below; currently it behaves like this:

1. when i set p3=0, it segfaults (because of instr Three, i think).
2. when i comment out the line with printf_i in instr Three, the 
segfault disappears.
3. when i set p3>0, the warnings about index out of range come up, and 
it segfaults.
4. when i comment out the line with printf_i in instr Three, the 
segfault disappears.

maybe the segfault has another reason, but maybe it is useful for you to 
test also with this more complex example.


<CsoundSynthesizer>
<CsOptions>
;-ndm0
</CsOptions>
<CsInstruments>
ksmps = 32

/****************************************************************************
S_Arr[], iLen StrToArr S_in, S_sep
Transforms the sections of the input string S_in to elements of a string 
array. The sections in S_in are seperated by the seperator S_in.

S_in - Input string.
S_sep - Seperator string.
S_Arr - Output string array.
iLen - Its length.
****************************************************************************/

opcode StrToArr, S[]i, SS

  S_in, S_sep xin

  ;count the number of substrings
  iLenSep strlen S_sep
  iPos = 0
  iPosShift = 0
  iCnt = 0

  while iPos != -1 do

   iCnt += 1
   S_sub strsub S_in, iPosShift
   iPos strindex S_sub, S_sep
   iPosShift += iPos+iLenSep

  od

  ;create a string array and put the substrings in it
  S_Arr[] init iCnt
  iPos = 0
  iPosShift = 0
  iArrIndx = -1
  while iPos != -1 do

   iArrIndx += 1
   S_sub strsub S_in, iPosShift
   iPos strindex S_sub, S_sep
   iEnd = (iPos == -1 ? -1 : iPosShift+iPos)
   S_ToArr strsub S_in, iPosShift, iEnd
   iPosShift += iPos+iLenSep
   S_Arr[iArrIndx] = S_ToArr

  od

  xout S_Arr, iCnt

endop


instr One

  S_InStr = 
"/my/file/one.wav|/my/file/two.wav|/my/file/three.wav|/my/file/four.wav"
  S_sepa = "|"
  S_arr[], iLen StrToArr S_InStr, S_sepa
  printf_i "One:\n  Input string = '%s'\n  Seperator = '%s'\n", 1, 
S_InStr, S_sepa

  indx = 0
  while indx < iLen do
   printf_i "  S_arr[%d] = '%s'\n", 1, indx, S_arr[indx]
   indx += 1
  od

endin

instr Two

  S_InStr = "/my/file/one.wav /my/file/two.wav /my/file/three.wav 
/my/file/four.wav"
  S_sepa = " "
  S_arr[], iLen StrToArr S_InStr, S_sepa
  printf_i "Two:\n  Input string = '%s'\n  Seperator = '%s'\n", 1, 
S_InStr, S_sepa

  indx = 0
  while indx < iLen do
   printf_i "  S_arr[%d] = '%s'\n", indx+1, indx, S_arr[indx]
   indx += 1
  od

endin

instr Three

  S_InStr = "/my/file/one.wav /my/file/two.wav /my/file/three.wav 
/my/file/four.wav"
  S_sepa = "/"
  S_arr[], iLen StrToArr S_InStr, S_sepa
  printf_i "Three:\n  Input string = '%s'\n  Seperator = '%s'\n", 1, 
S_InStr, S_sepa

  indx = 0
  while indx < iLen do
   printf_i "  S_arr[%d] = '%s'\n", indx+1, indx, S_arr[indx]
   indx += 1
  od

endin

</CsInstruments>
<CsScore>
i "One" 0 .1
i "Two" .1 .1
i "Three" .2 .1
</CsScore>
</CsoundSynthesizer>
Returns:
One:
   Input string = 
'/my/file/one.wav|/my/file/two.wav|/my/file/three.wav|/my/file/four.wav'
   Seperator = '|'
   S_arr[0] = '/my/file/one.wav'
   S_arr[1] = '/my/file/two.wav'
   S_arr[2] = '/my/file/three.wav'
   S_arr[3] = '/my/file/four.wav'
Two:
   Input string = '/my/file/one.wav /my/file/two.wav /my/file/three.wav 
/my/file/four.wav'
   Seperator = ' '
   S_arr[0] = '/my/file/one.wav'
   S_arr[1] = '/my/file/two.wav'
   S_arr[2] = '/my/file/three.wav'
   S_arr[3] = '/my/file/four.wav'
Three:
   Input string = '/my/file/one.wav /my/file/two.wav /my/file/three.wav 
/my/file/four.wav'
   Seperator = '/'
   S_arr[0] = ''
   S_arr[1] = 'my'
   S_arr[2] = 'file'
   S_arr[3] = 'one.wav '
   S_arr[4] = 'my'
   S_arr[5] = 'file'
   S_arr[6] = 'two.wav '
   S_arr[7] = 'my'
   S_arr[8] = 'file'
   S_arr[9] = 'three.wav '
   S_arr[10] = 'my'
   S_arr[11] = 'file'
   S_arr[12] = 'four.wav'

**Comment by jpffitch** [comment]
Believed fixed in 6.08 with new array rate computation
